integration/spark-common/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala

```diff
@@ -156,4 +161,206 @@ object DataLoadProcessBuilderOnSpark {
+   */
+  private def getSampleRDD(
+      sparkSession: SparkSession,
+      model: CarbonLoadModel,
+      hadoopConf: Configuration,
+      configuration: CarbonDataLoadConfiguration,
+      modelBroadcast: Broadcast[CarbonLoadModel]
+  ): RDD[(Object, Object)] = {
+    // initialize and prepare row counter
+    val configuration = DataLoadProcessBuilder.createConfiguration(model)
+    val header = configuration.getHeader
+    val rangeColumn = model.getRangePartitionColumn
+    val rangeColumnIndex = (0 until header.length).find{
+      index =>
+        header(index).equalsIgnoreCase(rangeColumn.getColName)
+    }.get
+    val rangeField = configuration
+      .getDataFields
+      .find(dataField => dataField.getColumn.getColName.equals(rangeColumn.getColName))
+      .get
+
+    // 1. Input
+    val newHadoopConf = new Configuration(hadoopConf)
+    newHadoopConf
+      .set(CSVInputFormat.SELECT_COLUMN_INDEX, "" + rangeColumnIndex)
+    val inputRDD = CsvRDDHelper
+      .csvFileScanRDD(sparkSession, model, newHadoopConf)
+      .mapPartitionsWithIndex { case (index, rows) =>
+        DataLoadProcessorStepOnSpark
+          .internalSampleInputFunc(rows, rangeField, index, modelBroadcast)
+      }
+
+    // 2. Convert
+    val conf = SparkSQLUtil.broadCastHadoopConf(sparkSession.sparkContext, hadoopConf)
+    val convertRDD = inputRDD
+      .mapPartitionsWithIndex { case (index, rows) =>
+        ThreadLocalSessionInfo.setConfigurationToCurrentThread(conf.value.value)
+        DataLoadProcessorStepOnSpark
+          .sampleConvertFunc(rows, rangeField, index, modelBroadcast)
+      }
+      .filter(_ != null)
+
+    convertRDD.map(row => (row.getObject(0), null))
+  }
+
+  /**
+   * calculate the number of partitions.
+   */
+  private def getNumPartitions(
+      configuration: CarbonDataLoadConfiguration,
+      model: CarbonLoadModel,
+      convertRDD: RDD[CarbonRow]
+  ): Int = {
+    var numPartitions = CarbonDataProcessorUtil.getGlobalSortPartitions(
+      configuration.getDataLoadProperty(CarbonCommonConstants.LOAD_GLOBAL_SORT_PARTITIONS))
+    if (numPartitions <= 0) {
+      if (model.getTotalSize <= 0) {
+        numPartitions = convertRDD.partitions.length
+      } else {
+        // calculate the number of partitions
+        // better to generate a CarbonData file for each partition
+        val totalSize = model.getTotalSize.toDouble
+        val table = model.getCarbonDataLoadSchema.getCarbonTable
+        val blockSize = 1024L * 1024 * table.getBlockSizeInMB
+        val blockletSize = 1024L * 1024 * table.getBlockletSizeInMB
+        val scaleFactor = if (model.getScaleFactor == 0) {
+          // here it assumes the compression ratio of CarbonData is about 30%,
+          // so it multiply by 3 to get the split size of CSV files.
+          3
+        } else {
+          model.getScaleFactor
+        }
+        val splitSize = Math.max(blockletSize, (blockSize - blockletSize)) * scaleFactor
```

Comment:
Can you add a comment what is background of doing this calculation this way?

Reply:
ok